### ISSUE TITLE
global variable destruction dependency problem

### ISSUE THREAD

**Issue by planetkr** [log4cplus/log4cplus#329]
Hi,
The logic under global-init.cxx::destroy_default_context_ will depend on this variable env.cxx::dir_sep, so when the global variable is released, if env.cxx::dir_sep is first released, it will result in global-init.cxx::destroy_default_context_  use the garbage value of this variable.

**Comment by wilx** [comment]
[**@planetkr**](https://github.com/planetkr) commented on [9 Aug 2018, 14:05 CEST](https://github.com/log4cplus/log4cplus/issues/329#issue-349098944 "2018-08-09T12:05:59Z - Replied by Github Reply Comments"):
> Hi,
> The logic under global-init.cxx::destroy_default_context_ will depend on this variable env.cxx::dir_sep, so when the global variable is released, if env.cxx::dir_sep is first released, it will result in global-init.cxx::destroy_default_context_  use the garbage value of this variable.

Why or how does `global-init.cxx::destroy_default_context_` depend on `dir_sep`?

**Comment by planetkr** [comment]
Open this option:
log4cplus.appender.TEST.CreateDirs=TRUET

And do not call  log4cplus::Logger::shutdown() when exit main().

**Comment by planetkr** [comment]
Yes, it's very hard to trigger this issue.  This issue was discovered by chance through the ASan test. ASan is a good tool, :)

**Comment by planetkr** [comment]
I have a little curious about (no offence), in patternlayout.cxx dir_sep defined like this:

static
log4cplus::tstring
get_basename (const log4cplus::tstring& filename)
{
#if defined(_WIN32)
    log4cplus::tchar const dir_sep(LOG4CPLUS_TEXT('\\\\'));
#else
    log4cplus::tchar const dir_sep(LOG4CPLUS_TEXT('/'));
#endif

    log4cplus::tstring::size_type pos = filename.rfind(dir_sep);
    if (pos != log4cplus::tstring::npos)
        return filename.substr(pos+1);
    else
        return filename;
}

why not define like  env.cxx::dir_rep do.

**Comment by FTSSYY** [comment]
I have met the same question.
